posts/2021-03-13-microprofile-config-2.0.adoc

```diff
@@ -0,0 +1,331 @@
+
+[source,java]
+----
+@Inject
+@ConfigProperty(name="server.url")
+String url; // retrieves the resolved value of “http://example.org”, or “http://example.test.org” if mp.config.profile is set to “testing”
+----
+
+You can also do some funky expressions, defining default values, composed expressions, and multiple expressions, which link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html#property-expressions[the spec] covers really well.
+
+Previous working configurations might behave differently if the configuration in use contains values with Property Expressions syntax.
+
+[#Extra-Info]
+== Some extra info for the return journey
+
+For the following examples, we'll use a slightly more rogue example ConfigSource (let's call it "example ConfigSource v2"):
+[[example-ConfigSource2]]
+.META-INF/microprofile-config.properties
+[source]
+----
+empty.property=
+empty.array.prop=,
+ports=9080,9081,9082
+server.port=9080
+----
+
+[#Config-Value-Behaviour-Updates]
+=== Empty And Special Values Behaviour Updates
+The behaviour for "empty" and "special" config property values has been updated:
+
+* The easiest way to get your head around this is to look at the link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html#_config_value_conversion_rules[conversion rule examples].
+* A value is considered to be "empty" if the link:https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/spi/Converter.html[Converter] being used considers it to be "empty". For example: 
+** All Converters consider `""`, the empty String, to be empty.
+** The built-in Converter for String[] considers `","` to be empty (because it is "special"). 
+* Retrieving these "empty" values "natively" will now throw a `NoSuchElementException`, e.g. for the values defined in the <<example-ConfigSource2, example ConfigSource v2>>:
++
+[source,java]
+----
+@Inject
+@ConfigProperty(name = "empty.property")
+String emptyProperty; // Will throw a `DeploymentException` caused by a `NoSuchElementException`
+
+@Inject
+@ConfigProperty(name = "empty.array.property")
+String[] emptyArrayProperty; // Will throw a `DeploymentException` caused by a `NoSuchElementException`
+----
++
+and
++
+[source,java]
+----
+Config config = ConfigProvider.getConfig();
+config.getValue("empty.property", String.class); // Will thorw a `NoSuchElementException`
+config.getValue("empty.array.property", String[].class); // Will thorw a `NoSuchElementException`
+----
++
+However these values can be retrieved "optionally":
++
+[source,java]
+----
+@Inject
+@ConfigProperty(name = "empty.property")
+Optional<String> emptyProperty; // returns "Optional.empty"
```

Comment:
I don't think you should have " (quotes) in `// returns "Optional.empty"` because it makes it look like it is returning a String